src/Microsoft.AspNetCore.Authorization.Policy/AuthenticationPolicyEvaluator.cs

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Authentication
+{
+    /// <summary>
+    /// Implements <see cref="IAuthenticationPolicyEvaluator"/>.
+    /// </summary>
+    public class AuthenticationPolicyEvaluator : IAuthenticationPolicyEvaluator
+    {
+        /// <summary>
+        /// Constructor.
+        /// </summary>
+        /// <param name="policyProvider">The The <see cref="IAuthenticationPolicyProvider"/>.</param>
+        public AuthenticationPolicyEvaluator(IAuthenticationPolicyProvider policyProvider)
+        {
+            Policies = policyProvider;
+        }
+
+        /// <summary>
+        /// Used to lookup AuthenticationPolicies.
+        /// </summary>
+        public IAuthenticationPolicyProvider Policies { get; }
+
+        /// <summary>
+        /// Authenticate for the specified authentication policy.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <param name="policyName">The name of the authentication policy.</param>
+        /// <returns>The result.</returns>
+        public virtual async Task<AuthenticateResult> AuthenticateAsync(HttpContext context, string policyName)
+        {
+            var policy = await Policies.GetAsync(context, policyName);
+            if (policy != null)
+            {
+                var count = policy.AuthenticateSchemes.Count();
+                if (count > 0)
+                {
+                    ClaimsPrincipal principal = null;
+                    foreach (var s in policy.AuthenticateSchemes)
+                    {
+                        var auth = await context.AuthenticateAsync(s);
```

Comment:
Isn't it cleaner to pass the IAuthenticationService to the ctor of this service then call `AuthenticateAsync` on that? Those extension methods are just for convenience...

Reply:
Yeah I wasn't sure which was better, I went with the context extensions since all the sugar overloads only exist on httpContext...